opm/grid/cpgrid/CpGrid.cpp

```diff
@@ -159,49 +162,87 @@ CpGrid::scatterGrid(EdgeWeightMethod method,
-        auto exportList = std::get<2>(part_and_wells);
-        auto importList = std::get<3>(part_and_wells);
+            std::tie(cell_part, wells_on_proc, exportList, importList)
+                = serialPartitioning
+                ? cpgrid::zoltanSerialGraphPartitionGridOnRoot(*this, wells, transmissibilities, cc, method, 0)
+                : cpgrid::zoltanGraphPartitionGridOnRoot(*this, wells, transmissibilities, cc, method, 0);
 #else
-        OPM_THROW(std::runtime_error, "Parallel runs depend on ZOLTAN. Please install!");
-        // std::vector<int> cell_part(current_view_data_->global_cell_.size());
-        // int  num_parts=-1;
-        // std::array<int, 3> initial_split;
-        // initial_split[1]=initial_split[2]=std::pow(cc.size(), 1.0/3.0);
-        // initial_split[0]=cc.size()/(initial_split[1]*initial_split[2]);
-        // partition(*this, initial_split, num_parts, cell_part, false, false);
-        // const auto& cpgdim =  logicalCartesianSize();
-        // std::vector<int> cartesian_to_compressed(cpgdim[0]*cpgdim[1]*cpgdim[2], -1);
-        // for( int i=0; i < numCells(); ++i )
-        // {
-        //     cartesian_to_compressed[globalCell()[i]] = i;
-        // }
-
-        // std::unordered_set<std::string> defunct_wells;
-
-        // if ( wells )
-        // {
-        //     cpgrid::WellConnections well_connections(*wells,
-        //                                              cpgdim,
-        //                                              cartesian_to_compressed);
-
-        //     auto wells_on_proc =
-        //         cpgrid::postProcessPartitioningForWells(cell_part,
-        //                                                 *wells,
-        //                                                 well_connections,
-        //                                                 cc.size());
-        //     defunct_wells = cpgrid::computeDefunctWellNames(wells_on_proc,
-        //                                                     *wells,
-        //                                                     cc,
-        //                                                     0);
-        // }
-#endif
+            OPM_THROW(std::runtime_error, "Parallel runs depend on ZOLTAN if useZoltan is true. Please install!");
+#endif // HAVE_ZOLTAN
+        }
+        else
+        {
+        std::vector<int> exportGlobalIds;
+        std::vector<int> exportLocalIds;
+        std::vector<int> exportToPart;
+        std::vector<int> importGlobalIds;
+        std::size_t numExport = 0;
+        int root = 0;
+
+        if (cc.rank() == root)
+        {
+            std::vector<int> parts(current_view_data_->global_cell_.size());
+            int  numParts=-1;
+            std::array<int, 3> initialSplit;
+            initialSplit[1]=initialSplit[2]=std::pow(cc.size(), 1.0/3.0);
+            initialSplit[0]=cc.size()/(initialSplit[1]*initialSplit[2]);
+            partition(*this, initialSplit, numParts, parts, false, false);
```

Comment:
In general, numParts will not be equal to cc.size(). First, the product of initialSplit[k], k = 0, 1, 2, may not be cc.size(). Second, numParts may become smaller than that product because of inactive cells (if the final argument was true, splitting to have connected partitions could also make it larger). It seems to me that the code subsequently assumes numParts == cc.size()?

Reply:
Actually `numParts`is not used below and hence disregarded. The only assumption is that `numParts <= cc.size()`. I guess that holds?
After the loadbalance all ranks will compute (possibly with an empty partition, which will result in an exception later on).

Reply:
> The only assumption is that `numParts <= cc.size()`. I guess that holds?

I think you are right, since we may only round down, not up, in lines 195 and 196. It may very easily result in empty partitions and termination, for example cc.size() == 9 will give just 8 partitions. However, since this is intended as a testing device, I think it is acceptable (if we want to offer this partitioning as a backup option for users we must revisit this).

You may merge when green.